### DISCUSSION TITLE
Deploying flet apps online

### DISCUSSION THREAD

**Discussion by kjurl** [flet-dev/flet#208] [Q&A] [answered]
I grabbed the code from [here](https://dev.to/watchakorn18k/create-app-with-flet-on-python-2oe7)
```python
import flet
from flet import IconButton, Page, Row, TextField, icons


def main(page: Page):
    page.title = "Flet counter example"
    page.vertical_alignment = "center"

    txt_number = TextField(value="0", text_align="right", width=100)

    def minus_click(e):
        txt_number.value = int(txt_number.value) - 1
        page.update()

    def plus_click(e):
        txt_number.value = int(txt_number.value) + 1
        page.update()

    page.add(
        Row(
            [
                IconButton(icons.REMOVE, on_click=minus_click),
                txt_number,
                IconButton(icons.ADD, on_click=plus_click),
            ],
            alignment="center",
        )
    )


flet.app(target=main, view=flet.WEB_BROWSER)
```

I was just wondering if I could deploy this on a cloud platform like heroku ( what should be the procfile ) ... it would be great if I could link it with a backend of my project ( fastapi + tortoise-orm ) ... thanks

**Comment by FeodorFitsner** [comment]
Hey, I've been looking into doing Heroku deployment guide, but then got distracted :)

Do you have experience with Heroku? If you bear with me we could go through the process together and do a guide for others! As far as I know they offer two deployment models at least: 1) git deployment and 2) Docker deployment. Which one are you thinking about?

**Reply by kjurl** [reply]
I'm guessing by git deployment you mean commiting a `Procfile` to your project. [This](https://devcenter.heroku.com/articles/procfile) page and [this](https://stackoverflow.com/questions/16128395/what-is-procfile-and-web-and-worker) question give a good explanation to Procfile. 

Basically, there are two process-types that we should be looking into `web` and `worker`. At my current knowledge I believe, flet should have a way to run the web-process ( `flet.app(target=main, view=flet.WEB_BROWSER)` ) through CLI then the Procfile can just be ( example ) :
```Procfile
web: flet web-run --target frontend.app:main --port $PORT
```
to run the app as a web process. ( I should mention as the app is highly inspired by flutter, I have no flutter experience ). Feel free to ask questions ^^

**Comment by FeodorFitsner** [comment]
OK, looking at https://devcenter.heroku.com/articles/procfile#the-web-process-type and https://stackoverflow.com/a/18670597/1435891 I guess we should have just `web: python main.py`.

Will give it a try tomorrow.

**Comment by FeodorFitsner** [comment]
But just to confirm you are looking into connecting your Heroku app to a GitHub repo, right?

![image](https://user-images.githubusercontent.com/5041459/185774577-cdfd413d-f6d4-4cdb-8a1b-3da833f99ab8.png)

**Reply by kjurl** [reply]
yes, that is the simplest workflow ^^
Edit: you can only view heroku logs through the heroku CLI `heroku logs --tail --app APP`, hope that helps

**Comment by FeodorFitsner** [comment]
Here you go: https://github.com/flet-dev/flet-heroku-app

**Comment by kjurl** [comment, marked as answer]
Cloning the repo as is and deploying it using the official python-buildpack gives an `Application Error`. These are the logs :
```
app[api]: Build succeeded
2022-08-21T19:33:49.477988+00:00 heroku[router]: at=error code=H14 desc="No web processes running" method=GET path="/" host=flet-heroku.herokuapp.com request_id=41e0a0e7-a302-4928-bb72-2f85f9cfdc65 fwd="132.154.238.49" dyno= connect= service= status=503 bytes= protocol=https
2022-08-21T19:33:49.922237+00:00 heroku[router]: at=error code=H14 desc="No web processes running" method=GET path="/favicon.ico" host=flet-heroku.herokuapp.com request_id=538a05e3-1ada-42a0-9a0f-b0e1f6d26c35 fwd="132.154.238.49" dyno= connect= service= status=503 bytes= protocol=https
```
Reading more into the logs [`H14 ERROR CODE`](https://devcenter.heroku.com/articles/error-codes#h14-no-web-dynos-running) refers to **No Web Dynos Running** 

Running: `heroku ps:scale web=1 --app <APP_NAME>` solved the issue for me

( I am leaving this here if anyone encounters this error in the future, and thanks for the amazing library ^^ )
